### DISCUSSION TITLE
[avram] Create a new empty `User` model object as User.new in ActiveRecord.

### DISCUSSION THREAD

**Discussion by zw963** [luckyframework/lucky#1712] [Q&A] [answered]
Hi, i am working on a new kemal + avram project which follow RESTful.

When deal with new/edit user page, both of them share same form.

so, i want to pass a user object to both new/edit page.

edit page is easy, i can retrieve exists user record use **UserQuery.find**

But, how to create a new not persisted user record? **User.new** is not work.

Thank you.

**Comment by jwoertink** [comment, marked as answer]
In Avram you can't have an empty model because this breaks the Lucky convention of type-safety first, and the representation that a model is just a read-only container of data. How you would do this would be to use a `SaveOperation`.

```crystal
get "/users/new" do |env|
  operation = SaveUser.new
end

post "/users" do |env|
  # I'm not totally sure how to write params from Kemal
  params = Avram::Params.new(env.params)
  SaveUser.create(params) do |op, user|

  end
end

get "/users/:user_id/edit" do |env|
   user = UserQuery.find(user_id)
   operation = SaveUser.new(user)
end

patch "/users/:user_id" do |env|
  params = Avram::Params.new(env.params)
  user = UserQuery.find(user_id)
  SaveUser.update(user, params) do |op, u|

  end
end

delete "/users/:user_id" do |env|
  user = UserQuery.find(user_id)
  DeleteUser.delete(user) do |op, u|

  end
end
```

Hope that helps.

**Reply by zw963** [reply]
Okay, i give a example for my usage, assume i will create/edit a company record.

```crsystal
    # render new page
    get path.admin_company_new do |env|
      errors = {} of Symbol => Array(String)
      company = SaveCompany.new

      render_admin "admin/companies/new.ecr"
    end

	# render edit page.
	get path.admin_company_edit do |env|
      errors = {} of Symbol => Array(String)

      company = CompanyQuery.find(env.params.url["id"])

      render_admin "admin/companies/edit.ecr"
    end
```

Both of them render page use same form.

1. When use with edit, pass a exist company use CompanyQuery is work, i can read company.name, company.code etc.. in view, to present the exists column value.

2. But, when use with new, `company = SaveCompany.new` not work.

```crystal
In src/views/partials/admin/company_form.ecr:4:83

 4 | <input id="company_name" type="text" name="company[name]", value="<%= company.name %>" />
                                                                                   ^---
Error: undefined method 'name' for Nil (compile-time type is (Company | Nil))
```

So, what i expected is just a `company` variable(empty record), make code like `company.name`, `company.code` all return empty string in view without complain error, for this use case, ` SaveCompany.new` not work as expect.

Thank you.

**Reply by jwoertink** [reply]
You should be able to do something like this:

```crystal
# render new page
get path.admin_company_new do |env|
  errors = {} of Symbol => Array(String)
  company = SaveCompany.new

  render_admin "admin/companies/new.ecr"
end

# render edit page.
get path.admin_company_edit do |env|
  errors = {} of Symbol => Array(String)

   company = SaveCompany.new(CompanyQuery.find(env.params.url["id"]))

   render_admin "admin/companies/edit.ecr"
end
```

**Reply by zw963** [reply]
> You should be able to do something like this:
> 
> ```crystal
> # render new page
> get path.admin_company_new do |env|
>   errors = {} of Symbol => Array(String)
>   company = SaveCompany.new
> 
>   render_admin "admin/companies/new.ecr"
> end
> 
> # render edit page.
> get path.admin_company_edit do |env|
>   errors = {} of Symbol => Array(String)
> 
>    company = SaveCompany.new(CompanyQuery.find(env.params.url["id"]))
> 
>    render_admin "admin/companies/edit.ecr"
> end
> ```

It works, but only part of, because when save/update failed, i need render new/edit view again, i have to pass same SaveCompany object again into view, following is my full example. i will explain the issue in the code.

```crystal
    # render edit page
    get path.admin_company_edit do |env|
      errors = {} of Symbol => Array(String)
      company = SaveCompany.new(CompanyQuery.find(env.params.url["company_id"]))

      render_admin "admin/companies/edit.ecr"
    end
```

render edit page works as expected, i can use `<%= company.name.value >` directly in my form.



```crystal	
	# Handle the edit page form submit.
    post path.admin_company_edit do |env|
	  # Issue 1: 
	  # I have to pass CompanyQuery instance here, because it must be pass as the first args of SaveCompany.update
      company = CompanyQuery.find(env.params.url["company_id"])
      params = {
        name: env.params.body["company[name]"].as(String)
      }

      SaveCompany.update(company, **params) do |operation, company|
        if operation.saved?
          env.redirect path.admin_company_index
        else
		  # Issue 2: 
		  # When update failed(assume name invalid), the block args company is nil, which cause <%= company.name.value > in form failed.
		  # In fact, i need a dirty company object here, with the invalid company name assign to object's name.
		  # so, will not lose the user input when rerender edit page again.
          errors = operation.errors
          render_admin "admin/companies/edit.ecr"
        end
      end
    end
```

Let me explain my issue.

1. Because i use same form for new/edit action, both of them share code like `<%= company.name.value %>` in ECR template.
   so, i expected action block pass same type company to view.
   
2. I expect a way(one object) to keep the user's output not lose when rerender the new/edit page again because validation issue.

Following is all my code.

```crystal
    # new record
    get path.admin_company_new do |env|
      errors = {} of Symbol => Array(String)
      company = SaveCompany.new

      render_admin "admin/companies/new.ecr"
    end
    post path.admin_company_new do |env|
      params = {
        name: env.params.body["company[name]"].as(String)
      }

      SaveCompany.create(**params) do |operation, _|
        if operation.saved?
          env.redirect path.admin_company_index
        else
          errors = operation.errors
          company = SaveCompany.new
          render_admin "admin/companies/new.ecr"
        end
      end
    end

    # edit a exist record
    get path.admin_company_edit do |env|
      errors = {} of Symbol => Array(String)
      company = SaveCompany.new(CompanyQuery.find(env.params.url["company_id"]))

      render_admin "admin/companies/edit.ecr"
    end
    post path.admin_company_edit do |env|
      company_object = CompanyQuery.find(env.params.url["company_id"])
      params = {
        name: env.params.body["company[name]"].as(String)
      }

      SaveCompany.update(company_object, **params) do |operation, _|
        if operation.saved?
          env.redirect path.admin_company_index
        else
          errors = operation.errors
          company = SaveCompany.new(company_object)
          render_admin "admin/companies/edit.ecr"
        end
      end
    end
```

it solve the issue 1, but as you see, it redundant, not good.

but for issue 2, it still not work, because i have to retrieve the fresh new object from db again, user input is lose.

Thank you.

**Reply by jwoertink** [reply]
I think what you would need to do is use the `operation` variable passed from the block argument... So something like this:

```crystal
SaveCompany.update(company_object, **params) do |operation, _|
    if operation.saved?
      env.redirect path.admin_company_index
    else
      errors = operation.errors
       # The `operation` here is the instance of SaveCompany.
       # It already knows about the company_object, and has all of the fields set.
       company = operation
      render_admin "admin/companies/edit.ecr"
    end
  end
```

Then on your form page, you can do `<%= company.name.errors %>` to display errors for that attribute, or use `company.errors` for all of them like you have set now.

**Reply by zw963** [reply]
Really cool, it works!  i never realized i can use operation instead.

In fact, i keep use name `operation` instead company or whatever directly in view, code is really simple for now.

Thank you very much.
